ksp_plugin/plugin.cpp

```diff
@@ -89,14 +101,37 @@ void Plugin::InsertCelestial(
       PlanetariumRotation().Inverse()(from_parent);
   LOG(INFO) << "In barycentric coordinates: " << relative;
   not_null<Celestial*> const celestial = inserted.first->second.get();
+  bodies_->emplace(celestial_index, std::move(body));
   celestial->set_parent(parent);
-  celestial->CreateHistoryAndForkProlongation(
-      current_time_,
-      parent->history().last().degrees_of_freedom() + relative);
+  DegreesOfFreedom<Barycentric> const& parent_degrees_of_freedom =
+      FindOrDie(*initial_state_, parent_index);
+  initial_state_->emplace(celestial_index,
+                          parent_degrees_of_freedom + relative);
 }
 
 void Plugin::EndInitialization() {
   initializing_.Flop();
+  std::vector<not_null<std::unique_ptr<MassiveBody const>>> bodies;
+  std::vector<DegreesOfFreedom<Barycentric>> initial_state;
+  for (auto& body : *bodies_) {
+    bodies.emplace_back(std::move(body.second));
+  }
+  bodies_.reset();
+  for (auto const& state : *initial_state_) {
+    initial_state.emplace_back(state.second);
+  }
+  initial_state_.reset();
+  ephemeris_ = std::make_unique<Ephemeris<Barycentric>>(std::move(bodies),
+                                                        initial_state,
+                                                        current_time_,
+                                                        history_integrator_,
+                                                        45 * Minute,
+                                                        1 * Milli(Metre));
+  for (auto const& index_celestial : celestials_) {
```

Comment:
In general, preference for pair as the loop variable, it makes it clear that we iterate over a map.